### PR TITLE
Nuget packages generation script for each project under src/ and some bugfixes

### PR DESCRIPTION
**Pull request by krwq**

### CODE REVIEW

**Comment by krwq** [comment]
I will be doing more testing with it today. I.e. if the output packages are actually consumable etc.. I'll create issues if there are any (quick look and it seems to be ok - package contains appropriate files)

File: packaging/nuget/package.ps1 [hunk 1]
```diff
@@ -0,0 +1,26 @@
+Push-Location $PSScriptRoot
+[Environment]::CurrentDirectory = $PWD
+
+$ProjectsDir = "..\..\src"
```
Comment (brthor) [inline packaging/nuget/package.ps1]: In other scripts we tend to import https://github.com/dotnet/cli/blob/master/scripts/_common.ps1 and make paths relative to the $RepoRoot variable

**Comment by brthor** [comment]
Will these nuget packages only work for windows? Seems like we should want linux packages as well.

**Comment by krwq** [comment]
AFAIK and was told dotnet pack generates bits for both Windows and Linux. Windows i the only place where we are planning for this to happen for now though.
@Sridhar-MS

File: packaging/nuget/package.ps1 (hunk 1 above)
Comment (krwq) [inline packaging/nuget/package.ps1]: What's the point of adding 2 lines of magic code which substitute 2 lines of obvious code?

**Comment by Sridhar-MS** [comment]
I think @brthor is asking why is there not an equivalent `package.sh` file.

**Comment by brthor** [comment]
@Sridhar-MS that's about right. Is there a reason we don't want x-plat packages?

**Comment by krwq** [comment]
We don't need creating identical nuget packages to happen on both linux and windows, I think one copy is enough

**Comment by brthor** [comment]
@krwq Are the packages x-plat? Is corehost included inside?

File: packaging/nuget/package.ps1 (hunk 1 above)
Comment (brthor) [inline packaging/nuget/package.ps1]: IMO, Consistency among our scripts and it makes reorganization easier.

**Comment by krwq** [comment]
We are not planning on publishing corehost for now. Publishing will be manual for the next few days and will happen only for selected binaries - the manual part is so we are sure that the packages contain what they should and there is no gaps before bot starts spamming a feed. After we gain confidence and consumers confirm stuff is working fine then we can move toward automating it and decide on how frequently should this happen.

The current approach is this:
- dotnet pack everything in src/
- for packages which are complete (have all bits) and we care about pushing (currently all of them are MSIL/xplat) push them manually to myget/nuget

I will work on auto-publishing separately which will filter out packages like corehost, currently the output folder will have nupkg for every project.json although only selected will be used when publishing.

**Comment by davidfowl** [comment]
We shouldn't be building packages for everything, just the libraries.

File: packaging/nuget/package.ps1 (hunk 1 above)
Comment (krwq) [inline packaging/nuget/package.ps1]: Tell me example of gain after changing this and I will

**Comment by krwq** [comment]
From what I heard we planned to package also commands/tools so that dotnet.exe can download them if they are missing. The only thing which I only considered excluding was template but since I noticed we can have nuget package for free it might actually be a good idea to replace dotnet init with nuget restore of that template.

File: packaging/nuget/package.ps1 (hunk 1 above)
Comment (brthor) [inline packaging/nuget/package.ps1]: Like I said I don't think this is a blocker, it's just the way paths are resolved in other scripts. If you prefer this way then go for it :)

One concern I do have though is what happens if one of the commands fails, and you exit before pop-location? Will that mess up other scripts?

File: packaging/nuget/package.ps1 [hunk 2]
```diff
@@ -0,0 +1,26 @@
+Push-Location $PSScriptRoot
+[Environment]::CurrentDirectory = $PWD
```
Comment (anurse) [inline packaging/nuget/package.ps1]: Is there a reason we aren't using `Convert-Path $PSScriptRoot`? It's more idiomatic PowerShell and does the same thing.

Comment (anurse) [inline packaging/nuget/package.ps1]: Also, changing the current directory is generally bad practice in a PowerShell script here. You'll need to wrap the whole thing in `try..finally` to ensure that if any thing fails in the script, the changes are reverted. I'd just use paths relative to either `$PSScriptRoot` or use `_common.ps1` and `$RepoRoot` everywhere.

**Comment by anurse** [comment]
We should definitely limit this to the actual libraries (ProjectModel, Workspaces, etc.) until we actually have a design for publishing and consuming tools from packages.

File: packaging/nuget/package.ps1 (hunk 1 above)
Comment (anurse) [inline packaging/nuget/package.ps1]: @brthor is right, the exit would occur before pop-location, see my comment above.

**Comment by krwq** [comment]
I don't see how any change in here would change amount of binaries produced. This is not being run as part of the build yet. AFAIK we are planning to have nuget packages for libraries and tools which makes it everything except dotnet-init template: for this particular case my first thought was to remove it and second thought that it is actually a good idea to make dotnet-init restore from nuget package. Where do I get the exact list of binaries which are needed? Who is responsible for design of this? I'd like to do this as part of this PR.

**Comment by davidfowl** [comment]
@krwq Because you can't use them and they don't make sense as nuget packages.

**Comment by piotrpMSFT** [comment]
Our core tools ship in the installers and should be grabbed that way. Extension tools, like test runners, will run from packages which is why we built that feature. Let's only pack the libraries in this repo at the moment as additional packages would muddy the story.

**Comment by krwq** [comment]
Ok will change

**Pull request merged**